[volume-6] 외부 시스템 장애 및 지연 대응

hyujikoh opened the pull request:
# 📌 Summary

**결제 시스템 Circuit Breaker 패턴 적용 및 타임아웃 전략 구현**

PG 결제 요청에 대한 안정성과 신뢰성을 향상시키기 위해 Resilience4j Circuit Breaker를 적용하고, 
500ms 타임아웃 기반의 비동기 처리 전략을 구현했습니다.

설정 및 구조 문서 : [resilience4j-implementation.md](https://github.com/hyujikoh/loopers-spring-java-template/pull/35/files#diff-5bc2cc5dad6a5c4ab01a679cb343b754b0f281e72804cd53ade3825c73cad7d4)


---

## ✅ Checklist

### **⚡ PG 연동 대응**

- [x]  PG 연동 API는 RestTemplate 혹은 FeignClient 로 외부 시스템을 호출한다.
- [x]  응답 지연에 대해 타임아웃을 설정하고, 실패 시 적절한 예외 처리 로직을 구현한다.
- [x]  결제 요청에 대한 실패 응답에 대해 적절한 시스템 연동을 진행한다.
- [x]  콜백 방식 + **결제 상태 확인 API**를 활용해 적절하게 시스템과 결제정보를 연동한다.

### **🛡 Resilience 설계**

- [x]  서킷 브레이커 혹은 재시도 정책을 적용하여 장애 확산을 방지한다.
- [x]  외부 시스템 장애 시에도 내부 시스템은 **정상적으로 응답**하도록 보호한다.
- [x]  콜백이 오지 않더라도, 일정 주기 혹은 수동 API 호출로 상태를 복구할 수 있다.
- [x]  PG 에 대한 요청이 타임아웃에 의해 실패되더라도 해당 결제건에 대한 정보를 확인하여 정상적으로 시스템에 반영한다.

### 테스트 완료
- [x] 단위 테스트: PaymentCircuitIntegrationTest (기존)
  - [x] minimumNumberOfCalls 검증
  - [x] failureRateThreshold 검증
  - [x] OPEN 상태 호출 차단
  - [x] HALF_OPEN 복구 검증
  - [x] Fallback 실행 검증
  - [x] 메트릭 검증

---
## 💬 Review Points
### 1. 외부 모듈 E2E 테스트의 적절한 범위
외부 PG 모듈과 통신하는 E2E 테스트를 어떻게 설계해야 할지 질문 드립니다.

PG Simulator와 통신하는 E2E 테스트(PaymentV1ApiE2ETest)를 작성했으나, 예상치 못한 케이스들이 많아서 제거했습니다.

현재 테스트 구조:

```
PaymentV1ApiE2ETest (E2E)
├── 실제 PG-Simulator와 HTTP 통신
├── 결제 요청 → PENDING 상태 저장
├── PG 콜백 수신 → SUCCESS/FAILED 처리
└── 타임아웃, 멱등성 등 검증

PaymentCircuitIntegrationTest (통합)
├── PgClient를 Mock으로 대체
├── Circuit Breaker 동작 검증
└── Retry, Fallback 로직 검증

```
문제점
1. 예측 불가능한 외부 모듈: PG Simulator의 응답이 예상과 다른 경우가 많음
2. 테스트 불안정성: 외부 모듈 의존으로 인한 Flaky Test 발생

제가 생각한 테스트 전략
1. 연결성 검증 (Smoke Test)
``` java
@Test
@DisplayName("PG 모듈 서버와 통신이 정상적으로 되는지 확인")
void pg_module_connection_test() {
    // Given: 간단한 Health Check 요청
    
    // When: PG 서버 호출
    
    // Then: HTTP 200 OK 응답
}
```
목적: PG 서버가 살아있고 통신 가능한지만 확인


2. 계약 검증 (Contract Test)
``` java
@Test
@DisplayName("PG 응답 구조가 정의된 스펙과 일치하는지 검증")
void pg_response_contract_test() {
    // Given: 결제 요청
    
    // When: PG 호출
    
    // Then: 
    // - meta.result 필드 존재
    // - data.transactionKey 필드 존재
    // - 필수 필드 타입 검증
}
```

목적: API 스펙 변경 감지 (Breaking Change 방지)

3. 부하 테스트를 통한 Circuit Breaker 검증

``` java
@Test
@DisplayName("실제 PG 장애 시 Circuit이 OPEN되는지 검증")
void circuit_breaker_with_real_pg_failure() {
    // Given: PG에 부하 발생 (타임아웃 유발)
    
    // When: 연속 호출
    
    // Then: Circuit OPEN 확인
}
```

목적: 실제 환경에서 Circuit Breaker 동작 확인



---

## 📎 References

### 관련 문서
- [Resilience4j Circuit Breaker 공식 문서](https://resilience4j.readme.io/docs/circuitbreaker)
- [Spring Cloud OpenFeign 공식 문서](https://docs.spring.io/spring-cloud-openfeign/docs/current/reference/html/)

---


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

# 릴리스 노트

## 새로운 기능
* 카드 결제 기능 추가 - 기존 포인트 결제와 함께 카드로 상품 결제 가능
* 결제 상태 추적 및 콜백 처리 - 결제 완료/실패/타임아웃 이벤트 처리
* 결제 시스템 안정성 개선 - 서킷브레이커 및 자동 재시도 메커니즘 적용

## 개선사항
* 주문 생성 프로세스 개선 - 결제 방식별 상세 정보 수집
* 비동기 결제 처리 - 논블로킹 콜백 핸들링

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hyujikoh committed:
feat(payment): 결제 API 및 관련 도메인 구현

- 결제 요청, 조회 API 추가
- 결제 상태 및 트랜잭션 정보 관리 로직 구현
- 예외 처리 및 응답 형식 정의

hyujikoh committed:
feat(payment): 결제 엔티티 및 상태, 타입 추가

hyujikoh committed:
feat(payment): 결제 엔티티에 필드 추가 및 수정

- 트랜잭션 키, 주문 번호, 금액, 콜백 URL, 거래 상태, 실패 사유, 완료 시간 필드 추가
- 카드 번호 필드 수정

hyujikoh committed:
feat(payment): 결제 엔티티에 필드 추가 및 수정

- 결제 요청 시각을 나타내는 requestedAt 필드 추가
- transactionStatus 필드를 nullable로 변경
- PgClient 클래스 추가
- build.gradle.kts에 OpenFeign 의존성 추가

hyujikoh committed:
Merge pull request #34 from hyujikoh/feat/failure-ready

PG 모듈 적용

hyujikoh committed:
feat(payment): PG 결제 클라이언트 및 설정 추가

- PG 결제 요청 및 정보 조회 API 구현
- 결제 요청 데이터 전송을 위한 DTO 추가
- PG 클라이언트 설정 및 재시도 정책 구성
- application.yml에 PG 시뮬레이터 URL 설정

hyujikoh committed:
feat(config): Feign 클라이언트 활성화

hyujikoh committed:
feat(config): Resilience4j 설정 추가

- Circuit Breaker, Retry, Timeout 설정을 추가하여
  pgClient의 안정성을 향상시킴
- 관련 라이브러리 의존성 추가

hyujikoh committed:
feat(payment): PG 클라이언트 로깅 레벨 설정 추가

hyujikoh committed:
feat(payment): 결제 클라이언트에 회로 차단기 및 재시도 설정 추가

hyujikoh committed:
feat(payment): 결제 리포지토리 및 예외 처리 추가

- PaymentJpaRepository, PaymentRepository, PaymentRepositoryImpl 추가
- PgClient의 fallbackFactory 구현
- 결제 서비스 장애 처리용 PgServiceUnavailableException 추가

hyujikoh committed:
feat(payment): 결제 관련 도메인 및 리포지토리 구현

- PaymentFacade, PaymentService, PaymentRepository, PaymentRepositoryImpl 추가
- PaymentJpaRepository에 주문 ID 및 트랜잭션 키로 조회하는 메서드 추가
- PG 클라이언트 관련 패키지 구조 변경

hyujikoh committed:
feat(payment): 결제 처리 로직 및 도메인 모델 추가

- PaymentCommand, PaymentDomainCreateRequest, PaymentEntity,
  PaymentInfo, PaymentFacade, PaymentService, PaymentV1ApiSpec,
  PaymentV1Controller, PaymentV1Dtos 클래스 구현
- 결제 요청 및 처리에 필요한 비즈니스 로직 추가
- Resilience4j를 활용한 회로 차단기 및 재시도 설정 적용

hyujikoh committed:
feat(payment): 결제 요청 처리 로직 개선

- PgPaymentRequest 생성 방식을 PaymentCommand를 사용하도록 변경
- PgClient 설정에서 fallbackFactory 제거
- TransactionStatus 열거형 추가

hyujikoh committed:
feat(payment): 결제 실패 처리 로직 추가

- 결제 실패 시 실패 사유를 기록하도록 PaymentEntity 수정
- 결제 요청 처리 시 실패 상태를 관리하는 메서드 추가
- PG 결제 요청 실패 시 결제 상태를 FAILED로 업데이트하는 로직 구현

hyujikoh committed:
feat(payment): 결제 상태 완료 처리 로직 추가

- 결제 완료 상태를 설정하는 complete 메서드 추가
- 결제 생성 팩토리 메서드 및 예외 처리 코드 포맷팅 개선

hyujikoh committed:
refactor(payment): 결제 생성 로직 리팩토링

- createPayment 메서드 제거 및 createFailedPayment 메서드 수정
- getByTransactionKey 메서드 위치 변경으로 코드 가독성 향상

hyujikoh committed:
feat(payment): 결제 타임아웃 처리 로직 추가

- PENDING 상태로 10분 이상 대기 중인 결제 건을 TIMEOUT 처리하는 스케줄러 구현
- 결제 실패 시 주문 취소 및 보상 처리 로직 추가
- 결제 상태에 따른 다양한 처리 메서드 추가

hyujikoh committed:
feat(payment): 결제 완료 및 실패 이벤트 리스너 추가

- 결제 완료 시 주문 확정 처리 로직 구현
- 결제 실패 시 주문 취소 처리 로직 구현
- 결제 이벤트 발행을 위한 PaymentCompletedEvent 및 PaymentFailedEvent 추가

hyujikoh committed:
feat(payment): 주문 ID 타입 Long으로 변경 및 패키지 구조 수정

- PaymentCompletedEvent 및 PaymentFailedEvent의 orderId 타입을 String에서 Long으로 변경
- 관련된 클래스의 패키지 경로를 application.payment에서 domain.payment.event로 수정
- OrderPaymentEventListener에서 orderId 처리 방식을 Long으로 변경하여 타입 일관성 확보

hyujikoh committed:
feat(order): 주문 생성 및 취소 로직 포인트 기반으로 변경

- 주문 생성, 취소, 확정 관련 메서드를 포인트 기반으로 수정
- 기존 createOrder 메서드를 createOrderByPoint로 변경하여 포인트 사용 로직 반영
- 주문 취소 시 포인트 환불 로직 추가
- 관련 테스트 코드 업데이트

hyujikoh committed:
feat(order): 카드 결제 주문 생성 로직 추가

- 결제 타입에 따라 주문 생성 로직 분기
- 카드 결제 시 포인트 차감 없이 주문만 생성하도록 구현
- 주문 정합성 검증 로직 추가

hyujikoh committed:
feat(order): 카드 결제 주문 생성 및 결제 요청 통합 처리

- 카드 결제 정보를 포함한 주문 생성 명령어 추가
- 카드 결제와 함께 주문을 생성하고 결제를 요청하는 메서드 구현
- 주문 생성 API에 카드 결제 관련 엔드포인트 추가

hyujikoh committed:
feat(order): 카드 결제 주문 생성 로직 리팩토링

- PaymentCommand와 PaymentInfo의 import 경로 간소화
- 주문 생성 및 결제 요청 API 엔드포인트 수정
- 불필요한 주석 제거 및 코드 정리

hyujikoh committed:
test(payment): PaymentEntity 단위 테스트 추가

- 결제 생성 및 상태 전이 관련 테스트 구현
- 유효한 정보로 PENDING 상태 결제 생성 검증
- 결제 요청이 null일 때 예외 발생 검증
- 결제 금액이 0 이하일 때 예외 발생 검증
- 결제 상태 전이 및 트랜잭션 키 업데이트 검증

hyujikoh committed:
feat(payment): 결제 콜백 처리 통합 테스트 추가

- PG-Simulator의 콜백 구조에 따른 결제 성공 및 실패 처리 테스트 구현
- 결제 상태 전이 및 이벤트 발행 검증
- 멱등성 검증 및 알 수 없는 상태 처리 테스트 추가
- 비동기 처리 대기 로직을 통해 데이터 일관성 확인

hyujikoh committed:
test(payment): 서킷 브레이커 동작 검증 통합 테스트 추가

- 서킷 브레이커의 최소 호출 수 및 실패율 임계값 검증
- OPEN 상태에서 호출 차단 및 Fallback 로직 실행 검증
- 메트릭 기록 정확성 확인

hyujikoh committed:
test(payment): 결제 처리 시 서킷 브레이커 동작 검증 테스트 추가

- 결제 처리 과정에서 Fallback이 정상적으로 작동하는지 검증
- Circuit Breaker 상태 및 메트릭 확인 로직 추가
- PaymentCommand 생성 시 사용자 정보를 포함하도록 수정

hyujikoh committed:
test(payment): 서킷 브레이커 상태 전환 및 복구 테스트 추가

hyujikoh committed:
test(payment): 서킷 브레이커 동작 검증 테스트 코드 개선

- 테스트 코드에서 중복된 import 문 제거
- given() 메서드 체이닝을 간소화하여 가독성 향상
- 주석 및 코드 정리를 통해 테스트 의도를 명확히 함

hyujikoh committed:
feat(payment): 결제 처리 로직 리팩토링 및 PG 응답 검증 추가

- 결제 처리 유스케이스를 PaymentProcessor로 분리하여 책임 명확화
- PG 결제 요청 및 응답 검증 로직을 PaymentValidator로 이동
- PG Gateway를 통한 결제 상태 조회 로직 추가
- 결제 콜백 처리 로직 개선 및 검증 강화

hyujikoh committed:
test(payment): 결제 콜백 처리 테스트 개선

- 랜덤 사용자명 생성 방식을 UUID로 변경하여 중복 가능성 감소
- PENDING 상태 응답을 위한 Mock 메서드 추가
- 알 수 없는 상태 처리 테스트 코드 제거

hyujikoh committed:
feat(payment): PG 클라이언트 타임아웃 및 재시도 설정 개선

- 연결 및 읽기 타임아웃을 300ms로 설정
- 재시도 최대 횟수를 1로 변경하여 타임아웃 시 재시도 비활성화
- 결제 처리 시 타임아웃 발생에 대한 예외 처리 로직 추가
- 테스트 클래스 이름 변경 및 관련 주석 수정

hyujikoh committed:
test(payment): 결제 회로 브레이커 테스트 문서 수정

hyujikoh committed:
test(payment): 결제 회로 브레이커 테스트 문서 수정

hyujikoh committed:
refactor(payment): 결제 콜백 처리 로직 개선

- 콜백 데이터 검증 주석 수정
- PG 콜백 처리 메서드 반환 타입을 Boolean으로 변경
- URI 상수를 사용하여 코드 가독성 향상

hyujikoh committed:
feat(payment): 결제 재시도 로직 및 페이징 기능 추가

- PG 클라이언트의 재시도 설정을 3회로 변경하고, 특정 예외에 대해서만 재시도하도록 수정
- 주문 항목 조회 시 페이징 기능을 추가하여 성능 개선
- 결제 완료 이벤트 및 실패 이벤트 처리 로직 수정

hyujikoh committed:
feat(order): 주문 번호 생성 및 유효성 검증 추가

- 주문 생성 요청에 주문 번호 필드 추가
- 주문 번호가 null 또는 빈 문자열일 경우 예외 처리
- 주문 번호 길이가 6자 미만일 경우 예외 처리
- 주문 엔티티에 주문 번호 필드 추가 및 생성 로직 수정

hyujikoh committed:
feat(payment): PG API 응답 실패 처리 로직 추가

- 결제 재시도 시 PG API 응답 실패에 대한 예외 처리 추가
- 주문 생성 및 결제 관련 클래스에서 주문 번호 필드 추가
- 결제 실패 시 보상 처리 로직 개선

hyujikoh committed:
feat(payment): 결제 API 타임아웃 설정 변경

- 연결 및 읽기 타임아웃을 300ms에서 500ms로 증가시켜
  안정성을 향상시킴
- 결제 정보에서 카드 종류를 "CREDIT"에서
  "SAMSUNG", "KB"로 변경하여 테스트 케이스 업데이트

hyujikoh committed:
feat(payment): 결제 e2e 테스트 제거

hyujikoh committed:
docs : 제거

hyujikoh committed:
feat(order): 주문 번호 Long 으로 변환

hyujikoh committed:
feat(payment): add error types for PG API failures and update validation logic

hyujikoh committed:
feat(payment): Resilience4j 기반 PG 통신 안정성 강화 문서 추가

- Circuit Breaker, Retry, Fallback 패턴 구현
- 장애 전파 방지 및 자동 복구 전략 설명
- E2E 테스트 및 통합 테스트 전략 포함

hyujikoh committed:
Merge pull request #35 from hyujikoh/feat/failure-ready

6주차 과제 구현

hyujikoh committed:
feat(payment): Resilience4j 기반 PG 통신 안정성 강화 문서 추가

- Circuit Breaker, Retry, Fallback 패턴 구현
- 장애 전파 방지 및 자동 복구 전략 설명
- E2E 테스트 및 통합 테스트 전략 포함

hyujikoh committed:
feat(payment): 결제 관련 DTO 및 유틸리티 개선

- 카드 번호 마스킹 유틸리티 추가
- 주문 생성 커맨드에 유효성 검증 추가
- 결제 명령 DTO에 카드 번호 마스킹 로직 추가
- 결제 결과 처리 로직 개선 및 이벤트 발행 추가

hyujikoh committed:
feat(payment): 결제 타임아웃 처리 로직 리팩토링

- 결제 타임아웃 처리 메서드를 분리하여 가독성 향상
- 트랜잭션 관리 개선으로 안정성 강화

hyujikoh committed:
feat(payment): 결제 타임아웃 처리 서비스 및 스케줄러 구현

- 결제 타임아웃을 처리하는 PaymentTimeoutProcessor 클래스 추가
- 스케줄러에서 PENDING 상태의 결제를 TIMEOUT 처리하는 로직 구현
- 각 결제 건마다 독립적인 트랜잭션으로 처리하여 부분 실패 허용

hyujikoh committed:
feat(payment): 결제 상태 처리 로직 개선

- PG 응답 상태를 문자열에서 enum으로 변환하는 메서드 추가
- 카드 번호 마스킹 로직 위치 변경으로 보안 강화

hyujikoh commented on apps/commerce-api/src/main/java/com/loopers/domain/payment/PaymentProcessor.java:
```diff
@@ -0,0 +1,93 @@
+
+import com.loopers.application.payment.PaymentCommand;
+import com.loopers.domain.user.UserEntity;
+import com.loopers.infrastructure.payment.client.dto.PgPaymentRequest;
+import com.loopers.infrastructure.payment.client.dto.PgPaymentResponse;
```
keep

hyujikoh commented on apps/commerce-api/src/main/java/com/loopers/domain/payment/PaymentValidator.java:
```diff
@@ -0,0 +1,135 @@
+            PaymentV1Dtos.PgCallbackRequest callbackRequest,
+            PgPaymentResponse pgData
+    ) {
+        // 주문 ID 검증
+        Long callbackOrderId = Long.parseLong(callbackRequest.orderId());
+        if (!payment.getOrderId().equals(callbackOrderId)) {
+            throw new IllegalArgumentException(
+                    String.format("주문 ID 불일치 - DB: %d, 콜백: %d",
+                            payment.getOrderId(), callbackOrderId)
+            );
+        }
```
keep

hyujikoh committed:
feat(payment): 결제 도메인 DTO 및 유틸리티 추가

hyujikoh committed:
refactor(payment): 결제 관련 주석 및 DTO 수정

- 결제 관련 클래스에서 DTO 타입을 명확히 지정
- 주석에서 불필요한 HTML 태그 제거
- 코드 가독성 향상 및 유지보수성 개선

hyujikoh committed:
refactor(payment): 결제 관련 주석 및 DTO 수정

- 결제 관련 클래스에서 DTO 타입을 명확히 지정
- 주석에서 불필요한 HTML 태그 제거
- 코드 가독성 향상 및 유지보수성 개선

hyujikoh committed:
Merge pull request #36 from hyujikoh/feat/failure-ready

Feat/failure ready

hyujikoh commented:
@coderabbitai full review

looper88888 commented:
⚡ PG 연동 대응
 PG 연동 API는 RestTemplate 혹은 FeignClient 로 외부 시스템을 호출한다.
 응답 지연에 대해 타임아웃을 설정하고, 실패 시 적절한 예외 처리 로직을 구현한다.
 결제 요청에 대한 실패 응답에 대해 적절한 시스템 연동을 진행한다.
 콜백 방식 + 결제 상태 확인 API를 활용해 적절하게 시스템과 결제정보를 연동한다.
🛡 Resilience 설계
 서킷 브레이커 혹은 재시도 정책을 적용하여 장애 확산을 방지한다.
 외부 시스템 장애 시에도 내부 시스템은 정상적으로 응답하도록 보호한다.
 콜백이 오지 않더라도, 일정 주기 혹은 수동 API 호출로 상태를 복구할 수 있다.
 PG 에 대한 요청이 타임아웃에 의해 실패되더라도 해당 결제건에 대한 정보를 확인하여 정상적으로 시스템에 반영한다.
 
 @coderabbitai 위 체크리스트에 맞게 구현되었는지 확인해줘

hyujikoh committed:
feat(payment): 결제 처리 로직에 회로 차단기 적용

- PG 결제 처리 메서드에 회로 차단기 추가
- 결제 요청 실패 시 Fallback 메서드 구현
- 결제 시스템 응답 지연 시 실패 결제 생성 로직 추가

hyujikoh committed:
feat(order): 주문 ID 관련 메서드 및 로그 수정

- 주문 ID를 orderNumber로 변경하여 일관성 유지
- 관련 메서드 및 로그 메시지 수정
- BrandEntity에 Builder 추가

hyujikoh committed:
feat(config): 실패 처리 대기 시간 및 예외 추가

- 실패 비율 임계값을 5초로 변경하여 빠른 회복 가능
- SocketTimeoutException 예외를 추가하여 타임아웃 처리 강화

hyujikoh committed:
refactor(product): 주석 포맷 통일 및 코드 정리

- 여러 클래스의 주석 포맷을 통일하여 가독성 향상
- 불필요한 import 문 제거 및 코드 정리

hyujikoh committed:
feat(payment): 결제 처리 로직에 재시도 기능 추가

- PG 결제 처리 시 실패 시 재시도 로직 구현
- 특정 예외 발생 시 최대 3회 재시도 설정
- 실패 시 결제 요청 처리 로직 개선

hyujikoh committed:
refactor(payment): 주문 번호 사용 일관성 개선

결제 처리 관련 테스트에서 주문 ID 대신 주문 번호를 사용하도록 수정하여
일관성을 높이고 가독성을 향상시킴

hyujikoh committed:
Merge pull request #37 from hyujikoh/feat/failure-ready

6주차 과제 기반 코드 리팩토링
